lib/rust/ensogl/core/src/debug/monitor.rs

```diff
@@ -931,3 +818,108 @@ stats_sampler!(
     0,
     1.0
 );
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use assert_approx_eq::assert_approx_eq;
+
+
+    struct TestSampler<S: Sampler> {
+        stats:   Stats,
+        sampler: S,
+        t:       f64,
+    }
+
+    impl<S: Sampler> TestSampler<S> {
+        fn new<F>(t0: f64, new_sampler: F) -> Self
+        where F: FnOnce(&Stats) -> S {
+            let stats = default();
+            let sampler = new_sampler(&stats);
+            Self { stats, sampler, t: t0 }
+        }
+    }
+
+    macro_rules! test_next_frame {
+        ($test:expr, $frame_time:expr, $post_frame_delay:expr, $expected_check:path, $expected_value:literal) => {
+            $test.stats.begin_frame($test.t);
+            $test.t += $frame_time;
+            $test.stats.end_frame($test.t);
+            assert_approx_eq!($test.sampler.value(), $expected_value, 0.001);
+            assert!(matches!($test.sampler.check(), $expected_check));
+            $test.t += $post_frame_delay;
+        };
+    }
```

Comment:
why this needs to be a macro and cannot be implemented just as a function?

Reply:
I started by making it a function (see: https://github.com/enso-org/enso/pull/3245/commits/58faf76bad15879ef10cc99a32281eba50cc41a6), but there was still some repetitiveness left due to `matches!` not being a function but a macro, and thus inability to pass just `$expected_check` as argument - the way I found around it was to pass a lambda, but it seemed rather clunky:

```rust
test.advance_frame(10.0, 6.0, |check| matches!(check, ValueCheck::Correct), 10.0);
```

Going with a macro doesn't seem to really be more complex, and the result seems easier on the eyes. I can bring back a function if you prefer?